site/src/components/pages/hub/BlockDataContainer.tsx

```diff
@@ -29,31 +38,81 @@ export const BlockDataContainer: VoidFunctionComponent<
   BlockDataContainerProps
 > = ({ metadata, schema, blockModule }) => {
   const [blockDataTab, setBlockDataTab] = useState(0);
-  const [blockTab, setBlockTab] = useState(0);
+  const [blockVariantsTab, setBlockVariantsTab] = useState(0);
   const [blockModalOpen, setBlockModalOpen] = useState(false);
+  const [alertSnackBarOpen, setAlertSnackBarOpen] = useState(false);
 
-  const example = metadata.examples?.[0];
+  const [text, setText] = useState("{}");
 
-  const [text, setText] = useState(
-    example ? JSON.stringify(example, undefined, 2) : "",
-  );
+  const previousBlockVariantsTab = useRef(-1);
+  const propertiesToRemove = useRef<string[]>([]);
 
   const theme = useTheme();
   const isMobile = useMediaQuery(theme.breakpoints.down("md"));
   const [activeMobileTab, setActiveMobileTab] = useState(0);
 
-  const prevPackage = useRef<string>(metadata.packagePath);
+  const prevPackage = useRef<string | undefined>(undefined);
+
   useEffect(() => {
-    if (prevPackage.current !== metadata.packagePath) {
+    if (prevPackage.current !== metadata?.packagePath) {
+      const example = {
+        ...metadata?.examples?.[0],
+        ...metadata?.variants?.[0].examples?.[0],
+        ...metadata?.variants?.[0].properties,
+      };
+
       // reset data source input when switching blocks
       if (example) {
         setText(JSON.stringify(example, undefined, 2));
       } else {
         setText("");
       }
+
+      prevPackage.current = metadata?.packagePath;
     }
-    prevPackage.current = metadata.packagePath;
-  }, [example, metadata.packagePath, text]);
+  }, [metadata, text]);
+
+  useEffect(() => {
+    if (
+      metadata?.variants &&
+      previousBlockVariantsTab.current !== blockVariantsTab
+    ) {
+      const blockVariant: BlockVariant | undefined =
+        metadata?.variants[blockVariantsTab];
+
+      try {
+        const parsedText = JSON.parse(text);
+
+        for (const propertyToRemove of propertiesToRemove.current) {
+          delete parsedText[propertyToRemove];
+        }
+
+        const nextText = {
+          ...blockVariant?.properties,
+          ...blockVariant.examples?.[0],
+          ...parsedText,
+        };
+
```

Comment:
I think we'd achieve the same goal if we do

```suggestion
        const nextText = {
            ...parsedText,
          ...blockVariant?.properties,
          ...blockVariant.examples?.[0],
        };
```

The necessary properties in `parsedText` will be overridden

Reply:
I guess this deserves a comment in the code. Some of the variants are also different by having a missing key in the `properties` field. Eg, all but the default variant in embed block have an `embedType` key. `embedType` prevents all embeds but the type specified in its value.

In cases like these, we would want the old keys to be removed. This is true for both the `examples` and `properties` field.

Reply:
In that case, then we probably don't need parsedText when switching between variants

```
variantA = { embedType: "twitter", url: "https://twitter.com/..."}
variantB = { embedType: "youtube", url: "https://youtube.com/..."}
```
Since both variants are mutually exclusive, we can just do

```
nextText = variantA 
nextText = variantB
```

instead of 
```
nextText = {
  ...variantA,
  ...variantB
}
```

Reply:
Thanks for adding this comment Teenoh. @CiaranMn this brings up an interesting point of discussion in this PR. Do we want the custom attributes added by the user to persist between variant switches?

If we only switch to the default example for the variant while ignoring user input, that simplifies the code significantly, but at the cost of losing a user's custom attributes at each variant switch.

Reply:
I don't think it's super important either way for the Block Hub demo. 

It's nice to keep the user's custom input, so long as the properties from the new variant consistently overwrite them (not the example - the variant properties itself). 

I'm not sure that the current code reliably does that - the only properties deleted from user input are from the _previous_ variant. If the next variant has some other property set, and the user has supplied their own value for it, the user's value will take precedence (since the parsed text is the last thing spread).

Reply:
Fixed this [here](https://github.com/blockprotocol/blockprotocol/pull/202/files#diff-aa262447e474cd1fbf253d93d617df697f68d9ccf4af71e9a5e1aa766bd9cd19R88) as discussed yesterday.